File: Array/TypeScript/_0566_ReshapetheMatrix.ts
function matrixReshape(mat: number[][], r: number, c: number): number[][] {
  if (r * c != mat.length * mat[0].length) return mat;

  let res: number[][] = [];
  let rowList: number[] = [];
  let row = 0;
  let col = 0;

  while (
    res.reduce((acc, cur) => (acc += cur.length), 0) <
    mat.length * mat[0].length
  ) {
    if (typeof mat?.[row]?.[col] === "undefined") {
      row += 1;
      col = 0;
    }
    rowList.push(mat?.[row]?.[col]);

    if (rowList.length == c) {
      res.push(rowList);
      rowList = [];
    }

    col += 1;
  }
  return res;
}

console.log(
  matrixReshape(
    [
      [1, 2],
      [3, 4],
    ],
    1,
    4
  )
);
console.log(
  matrixReshape(
    [
      [1, 2],
      [3, 4],
    ],
    2,
    4
  )
);
console.log(
  matrixReshape(
    [
      [1, 2],
      [3, 4],
    ],
    4,
    1
  )
);

File: Array/Python/_0566_ReshapetheMatrix.py
from typing import List


class Solution:
    # Iteration
    # Time Complexity: BigO(N^2)
    # Space Complexity: BigO(N^2)
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        if len(mat) * len(mat[0]) != r * c:
            return mat

        res = []
        rowlist = []

        for row in range(0, len(mat)):
            for col in range(0, len(mat[row])):
                rowlist.append(mat[row][col])

                if len(rowlist) == c:
                    res.append(rowlist)
                    rowlist = []

        return res


ins = Solution()

print(
    ins.matrixReshape(
        [
            [1, 2],
            [3, 4],
        ],
        1,
        4,
    )
)
print(
    ins.matrixReshape(
        [
            [1, 2],
            [3, 4],
        ],
        2,
        4,
    )
)
print(
    ins.matrixReshape(
        [
            [1, 2],
            [3, 4],
        ],
        4,
        1,
    )
)
